vsp.go

```diff
@@ -57,65 +61,325 @@ func (mw *MultiWallet) NewVSPClient(vspHost string, walletID int, purchaseAccoun
-	v.vspClient = newVSPClient(vspHost, nil, sourceWallet.Internal())
-	vspInfo, err := v.GetInfo(ctx)
+	vsp.ctx = ctx
+	vsp.client = newVSPClient(vspHost, nil, sourceWallet.Internal())
+	vspInfo, err := vsp.GetInfo(ctx)
 	if err != nil {
 		return nil, err
 	}
 
-	v.ctx = ctx
-	v.vspClient.pub = vspInfo.PubKey
-	return v, nil
+	vsp.client.pub = vspInfo.PubKey
+	return vsp, nil
 }
 
-func (v *VSP) PoolFee(ctx context.Context) (float64, error) {
+//  PoolFee returns the vsp fee percentage.
+func (vsp *VSP) PoolFee(ctx context.Context) (float64, error) {
 	var vspInfo VspInfoResponse
-	err := v.vspClient.get(ctx, apiVSPInfo, &vspInfo)
+	err := vsp.client.get(ctx, apiVSPInfo, &vspInfo)
 	if err != nil {
 		return -1, err
 	}
 	return vspInfo.FeePercentage, nil
 }
 
 // PurchaseStakeTickets purchases the number of tickets passed as an argument and pays the fees for the tickets
-func (v *VSP) PurchaseTickets(ticketCount, expiryBlocks int32, passphrase []byte) error {
-	err := v.w.UnlockWallet(passphrase)
+func (vsp *VSP) PurchaseTickets(ticketCount, expiryBlocks int32, passphrase []byte) error {
+	err := vsp.w.UnlockWallet(passphrase)
 	if err != nil {
 		return translateError(err)
 	}
-	defer v.w.LockWallet()
+	defer vsp.w.LockWallet()
 
 	request := &w.PurchaseTicketsRequest{
 		Count:                int(ticketCount),
-		SourceAccount:        v.purchaseAccount,
-		Expiry:               v.w.GetBestBlock() + expiryBlocks,
+		SourceAccount:        vsp.purchaseAccount,
+		Expiry:               vsp.w.GetBestBlock() + expiryBlocks,
 		MinConf:              DefaultRequiredConfirmations,
-		VSPFeeProcess:        v.PoolFee,
-		VSPFeePaymentProcess: v.ProcessFee,
+		VSPFeeProcess:        vsp.PoolFee,
+		VSPFeePaymentProcess: vsp.ProcessFee,
 	}
 
-	networkBackend, err := v.w.Internal().NetworkBackend()
+	networkBackend, err := vsp.w.Internal().NetworkBackend()
 	if err != nil {
 		return err
 	}
 
-	_, err = v.w.Internal().PurchaseTickets(v.ctx, networkBackend, request)
+	_, err = vsp.w.Internal().PurchaseTickets(vsp.ctx, networkBackend, request)
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
 
+// StartTicketBuyer starts the automatic ticket buyer.
+func (vsp *VSP) StartTicketBuyer(balanceToMaintain int64, passphrase []byte) error {
+	if balanceToMaintain < 0 {
+		return errors.New("Negative balance to maintain given")
+	}
+
+	vsp.w.cancelAutoTicketBuyerMux.Lock()
+	cancelFunc := vsp.w.cancelAutoTicketBuyer
+	vsp.w.cancelAutoTicketBuyerMux.Unlock()
+	if cancelFunc != nil {
+		return errors.New("Ticket buyer already running")
+	}
+
+	ctx, cancel := vsp.w.shutdownContextWithCancel()
+
+	vsp.w.cancelAutoTicketBuyerMux.Lock()
+	vsp.w.cancelAutoTicketBuyer = cancel
+	vsp.w.cancelAutoTicketBuyerMux.Unlock()
+
+	if len(passphrase) > 0 && vsp.w.IsLocked() {
+		err := vsp.w.UnlockWallet(passphrase)
+		if err != nil {
+			return translateError(err)
+		}
+		defer vsp.w.LockWallet()
+	}
+
+	go func() {
+		log.Infof("Running ticket buyer on Wallet: %s", vsp.w.Name)
```

Comment:
```suggestion
		log.Infof("Running ticket buyer on wallet: %s", vsp.w.Name)
```
Consider using consistent logging patterns for the ticket buyer. Other logging use `[%d]` and walletID.